src/msw/thread.cpp

```diff
@@ -1121,6 +1122,80 @@ wxThreadError wxThread::Kill()
     return rc;
 }
 
+// This function requires at least MSVC 2017 version 15.6 and
+// Windows 10 version 1607 to succeed.
+static void wxSetThreadNameOnWindows10(const wxChar *threadName)
+{
+    typedef HRESULT(WINAPI* SetThreadDescription_t)(HANDLE, PCWSTR);
+    static SetThreadDescription_t s_pfnSetThreadDescription = NULL;
+
+    static bool s_initDone = false;
+    if ( !s_initDone )
+    {
+        wxLoadedDLL dllKernel32("kernel32.dll");
+        wxDL_INIT_FUNC(s_pfn, SetThreadDescription, dllKernel32);
+        s_initDone = true;
+    }
+
+    if ( s_pfnSetThreadDescription ) {
+        if ( FAILED((*s_pfnSetThreadDescription)(
+            GetCurrentThread(),
+            threadName)) )
+        {
+            wxLogSysError("SetThreadDescription failed");
+        }
+    }
+}
+
+// This function works with all MSVC versions.
+static void wxSetThreadNameOnAnyMSVC(const char* threadName)
+{
+    // This implementation is taken almost verbatim from:
+    // https://msdn.microsoft.com/en-us/library/xcb2z8hs.aspx
```

Comment:
> Not sure what are the licensing implications of this :-(

If the piece of code is long enough to be copyrightable, reformatting and refactoring it doesn't change the situation much.

However, Microsoft's docs appear to be hosted on Github nowadays, and all the code samples are licensed under the MIT license: https://github.com/MicrosoftDocs/visualstudio-docs/blob/master/ThirdPartyNotices

> Looking at it again, my version is probably broken in 64 bits, as it uses `DWORD` instead of `ULONG_PTR`. And if your version works (I assume you have tested it in 64 bits?), this must be wrong.

Yes, my code works on 64 bits. I tried replacing `sizeof(ULONG_PTR)` with `sizeof(DWORD)` and it still seems to work.